File: packages/core/cognis/contract.py
"""Single source of truth for the extension ↔ backend contract version.

The VS Code extension and the Python backend (``cognis-cli``, ``cognis-indexd``,
``cognis-mcpd``) exchange JSON across process and language boundaries. Those
shapes are a cross-language contract: if one side renames or drops a field, or
the set of commands / MCP tools changes, the other side breaks — and because
the extension updates independently of the backend (marketplace vs PyPI), a
version-skewed install is the single most common production state and the one
the matched-version e2e suite never exercises.

``CONTRACT_VERSION`` is the explicit handshake token. Bump it whenever the shape
of the cross-process exchange changes in a way the other side must know about
(a renamed/removed field, a new required command, a changed MCP tool surface).
The extension ships the version it was built against, calls ``cognis-cli
handshake`` at startup, and degrades with a clear, actionable message instead of
failing silently when the versions disagree.
"""

from __future__ import annotations

from typing import Any, Final

from cognis import __version__

#: Bump on any breaking change to the extension ↔ backend JSON contract.
CONTRACT_VERSION: Final[int] = 1

#: CLI commands the extension relies on (see cli.ts / workspace.ts). The
#: handshake advertises these so the extension can detect a backend that lacks a
#: command it needs.
CLI_COMMANDS: Final[tuple[str, ...]] = (
    "init",
    "bootstrap",
    "health",
    "paths",
    "doctor",
    "mcp-config",
    "index",
    "handshake",
)

#: MCP tools the server exposes (see cognis_mcpd/server.py). The AI agent calls
#: these; pinned here so a dropped/renamed tool is caught by a contract test and
#: surfaced in the handshake capability list.
MCP_TOOLS: Final[tuple[str, ...]] = (
    "diffuse_context",
    "symbol_lookup",
    "symbol_search",
    "discover_symbols",
    "semantic_search",
    "resolve_symbols",
    "dependency_trace",
    "retrieve_context_capsule",
)


def handshake_payload() -> dict[str, Any]:
    """Build the ``cognis-cli handshake`` JSON the extension reads at startup."""
    return {
        "contract_version": CONTRACT_VERSION,
        "engine_version": __version__,
        "cli_commands": list(CLI_COMMANDS),
        "mcp_tools": list(MCP_TOOLS),
    }

File: crates/cognis-core/src/contract.rs
//! Extension ↔ backend contract — Rust mirror of `packages/core/cognis/contract.py`.
//!
//! Single source of truth for the cross-process JSON contract. Order and values
//! are kept identical to the Python contract so the existing TypeScript
//! extension keeps working unchanged (Requirements 3.1, 3.3).

use serde::Serialize;

/// Bump on any breaking change to the extension ↔ backend JSON contract.
pub const CONTRACT_VERSION: u32 = 1;

/// CLI commands the extension relies on (same order as `contract.py`).
pub const CLI_COMMANDS: [&str; 8] = [
    "init",
    "bootstrap",
    "health",
    "paths",
    "doctor",
    "mcp-config",
    "index",
    "handshake",
];

/// MCP tools the server exposes (same order as `contract.py`).
pub const MCP_TOOLS: [&str; 8] = [
    "diffuse_context",
    "symbol_lookup",
    "symbol_search",
    "discover_symbols",
    "semantic_search",
    "resolve_symbols",
    "dependency_trace",
    "retrieve_context_capsule",
];

/// The `cognis-cli handshake` payload the extension reads at startup.
#[derive(Debug, Clone, Serialize)]
pub struct HandshakePayload {
    pub contract_version: u32,
    pub engine_version: String,
    pub cli_commands: Vec<String>,
    pub mcp_tools: Vec<String>,
}

/// Build the handshake payload (mirrors `handshake_payload()`).
pub fn handshake_payload() -> HandshakePayload {
    HandshakePayload {
        contract_version: CONTRACT_VERSION,
        engine_version: env!("CARGO_PKG_VERSION").to_string(),
        cli_commands: CLI_COMMANDS.iter().map(|s| s.to_string()).collect(),
        mcp_tools: MCP_TOOLS.iter().map(|s| s.to_string()).collect(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tool_set_and_order_match_contract() {
        assert_eq!(MCP_TOOLS.len(), 8);
        assert_eq!(MCP_TOOLS[0], "diffuse_context");
        assert_eq!(MCP_TOOLS[1], "symbol_lookup");
        assert_eq!(MCP_TOOLS[7], "retrieve_context_capsule");
        assert_eq!(CLI_COMMANDS[0], "init");
        assert_eq!(CLI_COMMANDS[7], "handshake");
    }

    #[test]
    fn handshake_payload_serializes() {
        let p = handshake_payload();
        let v = serde_json::to_value(&p).unwrap();
        assert_eq!(v["contract_version"], 1);
        assert_eq!(v["mcp_tools"].as_array().unwrap().len(), 8);
        assert_eq!(v["cli_commands"][0], "init");
    }
}
